src/hotspot/share/opto/escape.cpp

```diff
@@ -401,6 +430,329 @@ bool ConnectionGraph::compute_escape() {
+                  add_edge(ptnode_adr(new_load->_idx), ptnode_adr(new_addp->_idx)->as_Field());
+                }
+              }
+            }
+          }
+        }
+        --k;
+        k = MIN2(k, (int)previous_addp->outcnt()-1);
+      }
+
+      // Remove the old AddP from the processing list because it's dead now
+      alloc_worklist.remove_if_existing(previous_addp);
+    }
+    j -= num_edges;
+    j = MIN2(j, (int)ophi->outcnt()-1);
+  }
+}
+
+// This method will create a SafePointScalarObjectNode for each combination of
+// scalar replaceable allocation in 'ophi' and SafePoint node in 'safepoints'.
+// Each SafePointScalarObjectNode created here may describe multiple scalar
+// replaced objects - check detailed description in SafePointScalarObjectNode
+// class header.
+//
+// This method will set entries in the Phi that are scalar replaceable to 'null'.
+void ConnectionGraph::reduce_this_phi_on_safepoints(PhiNode* ophi, Unique_Node_List* safepoints) {
+  Node* minus_one           = _igvn->register_new_node_with_optimizer(ConINode::make(-1));
+  Node* selector            = _igvn->register_new_node_with_optimizer(PhiNode::make(ophi->region(), minus_one, TypeInt::INT));
+  Node* null_ptr            = _igvn->makecon(TypePtr::NULL_PTR);
+  const TypeOopPtr* merge_t = _igvn->type(ophi)->make_oopptr();
+  uint number_of_sr_objects = 0;
+
+  _igvn->hash_delete(ophi);
+
+  // Fill in the 'selector' Phi. If index 'i' of the selector is:
+  // -> a '-1' constant, the i'th input of the original Phi is NSR.
+  // -> a 'x' constant >=0, the i'th input of of original Phi will be SR and the
+  //    info about the scalarized object will be at index x of
+  //    ObjectMergeValue::possible_objects
+  for (uint i = 1; i < ophi->req(); i++) {
+    Node* base          = ophi->in(i);
+    JavaObjectNode* ptn = unique_java_object(base);
+
+    if (ptn != nullptr && ptn->scalar_replaceable()) {
+      Node* sr_obj_idx = _igvn->register_new_node_with_optimizer(ConINode::make(number_of_sr_objects));
+      selector->set_req(i, sr_obj_idx);
+      number_of_sr_objects++;
+    }
+  }
+
+  // Update the debug information of all safepoints in turn
+  for (uint spi = 0; spi < safepoints->size(); spi++) {
+    Node* call      = safepoints->at(spi);
+    Node* ctrl      = call->in(TypeFunc::Control);
+    Node* memory    = call->in(TypeFunc::Memory);
+    JVMState *jvms  = call->jvms();
+    uint merge_idx  = (call->req() - jvms->scloff());
+    int debug_start = jvms->debug_start();
+
+    SafePointScalarMergeNode* smerge = new SafePointScalarMergeNode(merge_t, merge_idx);
+    smerge->init_req(0, _compile->root());
+    _igvn->register_new_node_with_optimizer(smerge);
+
+    // Keep a copy of the original pointer to NSR objects
+    call->add_req(ophi);
+
+    // Add the selector so we know which direction the execution took
+    call->add_req(selector);
+
+    for (uint i = 1; i < ophi->req(); i++) {
```

Comment:
Comparing to new_phi and selector, I think this is the heavy-lifting work. You "replace" all appearances of ptn with SPSO. This logic almost overlaps the 'scalar replacement' part in MacroExpand. 

Do you consider to perform the transformation in MacroExpand? Your prior changes have already removed NSR marks, ME/SR will consider 'ptn'.